Shadow Memory: Refactoring, enabling tests and minor doxygen improvements.

NlightNFotis opened the pull request:
The second part of our work on the `set_field`, this is enabling some tests that are now passing after our bug fixes, refactoring code to use a more modern style at places, and adding some doxygen documentation for some of the functions.

This is really just a checkin of our current work in progress before we move forward with more tests && refactors.

Work done with @esteffin

peterschrammel committed:
Correctly set value-set dereference pointer

NlightNFotis committed:
Shadow memory implementation refactor and doxygen

Commit:
Fix SAT failure when bit-or-ing pointers

When setting shadow memory for pointer types the value is replicated by
applying left shift and combining the shifted values with bitwise-or.
However the SAT solver does not support bitwise-or for pointer types.

This commit changes the way the replication is done for pointer types by
converting them to an unsigned bitvector of the pointer type size,
performing the shift and bitwise-or on this bitvector type and then
casting it to pointer at the end.

NlightNFotis committed:
Marking CORE more Shadow Memory regression tests

Marking as CORE all the shadow memory regression tests that are now
passing.

The remaining regression tests that are still marked as KNOWNBUG will be
added as soon as the implementation will be fixed.

peterschrammel commented on src/goto-symex/shadow_memory_util.cpp:
```diff
@@ -691,7 +690,17 @@ static exprt get_matched_expr_cond(
 }
 
-// TODO: doxygen?
+// TODO: FOtis - revisit.
```
Docs lgtm.

peterschrammel commented on src/goto-symex/shadow_memory_util.cpp:
```diff
@@ -22,7 +22,6 @@ Author: Peter Schrammel
 #include <util/ssa_expr.h>
 #include <util/std_expr.h>
 
-
```
This commit should be squashed into the respective previous commits.

esteffin replied:
Agreed.

Changed

Commit:
Fix missing alloca definition on regression test

Windows requires alloca to be defined otherwise an invariant will be
violated (the invariant violation will be changed in another subsequent
PR).
